scripts/import-cp-contexts.sh

```diff
@@ -0,0 +1,249 @@
+    fi
+elif [[ "$KUBECONFIG" != "" ]] ; then
+    in_KUBECONFIG="$KUBECONFIG"
+    if [[ "$arg_output" == "" ]] ; then
+        if [[ "$KUBECONFIG" == *":"* ]] ; then
+            # hard to decide may be out_kubeconfig="$HOME/.kube/config"
+            >&2 echo "ERROR: please specify an output kubeconfig filename with \"-o <filename>\"!"
+            exit 1
+        else
+            out_kubeconfig="$KUBECONFIG"
+        fi
+    fi
+else
+    in_KUBECONFIG="$HOME/.kube/config"
+    if [[ "$arg_output" == "" ]] ; then
+        out_kubeconfig="$HOME/.kube/config"
+    else
+        out_kubeconfig="$arg_output"
+    fi
+fi
+echov "- input kubeconfigs: \"$in_KUBECONFIG\""
+echov "- output kubeconfig: \"$out_kubeconfig\""
+
+
+# Determine the list of contexts to search
+echov "Determining the list of contexts to search:"
+if [[ "$arg_context" == "" ]] ; then
+    contexts=($(KUBECONFIG="$in_KUBECONFIG" kubectl config get-contexts --no-headers -o name))
+else
+    contexts=("$arg_context")
+fi
+for context in "${contexts[@]}" ; do # for all contexts
+    echov "- found context \"${context}\""
+done
+
+
+echov "Getting kubeconfigs of KubeFlex Control Planes:"
+n=0
+for context in "${contexts[@]}" ; do # for all contexts
+    echov "- searching context \"${context}\""
+    cps=($(KUBECONFIG="$in_KUBECONFIG" kubectl --context "${context}" get controlplanes -no-headers -o name 2> /dev/null || true))
+    for i in "${!cps[@]}" ; do # for all control planes in context ${context}
+        name=${cps[i]##*/}
+        if [[ "$arg_names" != "" && ",$arg_names," != *",$name,"* ]] ; then
+            echov "  - skipping KubeFlex Control Plane \"$name\""
+            continue
+        fi
+        cp_name[n]=$name
+        cp_type[n]=$(KUBECONFIG="$in_KUBECONFIG" kubectl get controlplane ${cp_name[n]} -o jsonpath='{.spec.type}')
+        echov "  - found KubeFlex Control Plane \"${cp_name[i]}\" of type \"${cp_type[i]}\""
+        cp_kubeconfig[n]=$(get_kubeconfig "${context}" "${cp_name[n]}" "${cp_type[n]}" "$arg_ip_addr")
+        n=$((n+1))
+    done
+done
+
+
+if [[ "${#cp_name[@]}" == "0" ]] ; then
+    echov "No KubeFlex Control Planes found... nothing to do"
+    exit 0
+fi
+
+
+echov "Merging the contexts of KubeFlex Control Planes:"
+kubeconfig_list=""
+for i in "${!cp_name[@]}" ; do
+    echov "- \"${cp_name[i]}\" of type \"${cp_type[i]}\" ==> saving to temporary file \"kubeconfig_${cp_name[i]}\""
+    echo "${cp_kubeconfig[i]}" > "kubeconfig_${cp_name[i]}"
+    kubeconfig_list="$kubeconfig_list:kubeconfig_${cp_name[i]}"
+done
+if [[ "$arg_merge" == "true" ]] ; then
+    echov "* including current kubeconfig: \"$in_KUBECONFIG\""
+    merge_KUBECONFIG="$in_KUBECONFIG$kubeconfig_list"
+else
+    merge_KUBECONFIG="${kubeconfig_list:1}"
+fi
+if [[ "$out_kubeconfig" == "-" ]] ; then
+    KUBECONFIG="$merge_KUBECONFIG" kubectl config view --flatten
+else
+    (KUBECONFIG="$merge_KUBECONFIG" kubectl config view --flatten) > "kubeconfig_tmp"
```

Comment:
A given control plane name can be found in multiple input contexts. That will lead to collisions here. What happens in this case? The answer should at least be document, and preferably be a defensible outcome.

Reply:
I'm not sure

Reply:
I tested, using commit e8f1666eb1614408654cfa1cbbaa14a0eb7f04f6 and my existing default kubeconfig. It went like this:

```
mspreitz@mjs13 kubestellar % scripts/import-cp-contexts.sh -o test.kubeconfig          
Determining the list of kubeconfigs to search:
- input kubeconfigs: "/Users/mspreitz/.kube/config"
- output kubeconfig: "test.kubeconfig"
Determining the list of contexts to search:
...
Merging the contexts of KubeFlex Control Planes:
- "its1" of type "vcluster" ==> saving to temporary file "kubeconfig_its1"
- "wds1" of type "k8s" ==> saving to temporary file "kubeconfig_wds1"
- "its1" of type "vcluster" ==> saving to temporary file "kubeconfig_its1"
- "wds1" of type "k8s" ==> saving to temporary file "kubeconfig_wds1"
- "its1" of type "vcluster" ==> saving to temporary file "kubeconfig_its1"
- "wds1" of type "k8s" ==> saving to temporary file "kubeconfig_wds1"
- "its1" of type "vcluster" ==> saving to temporary file "kubeconfig_its1"
- "wds1" of type "k8s" ==> saving to temporary file "kubeconfig_wds1"
* backing up "test.kubeconfig" to "test.kubeconfig.bak"
```

And then:

```
mspreitz@mjs13 kubestellar % ls kubeconfig_*
kubeconfig_its1	kubeconfig_tmp	kubeconfig_wds1

mspreitz@mjs13 kubestellar % open -a Emacs test.kubeconfig
The file /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/test.kubeconfig does not exist.

mspreitz@mjs13 kubestellar % ls -ld test.*
zsh: no matches found: test.*
```

Reply:
Yes, moving a non-existent file fails.

```
mspreitz@mjs13 kubestellar % mv -f 404 404.bak
mv: rename 404 to 404.bak: No such file or directory
mspreitz@mjs13 kubestellar % echo $?
1
```

Reply:
My original complaint here was mistaken; even if a given ControlPlane name is found in multiple contexts, only one `kubeconfig_$cp_name` is created and there is no collision here.

The behavior in this case needs to be documented, and the `mv` problem needs to be fixed.

Reply:
I fixed the error on the move

Reply:
having multiple cp with the same name in different contexts or kubeconfigs is not very likely, in my opinion

Reply:
When I tested this script, it found four contexts in my `~/.kube/config` that have a ControlPlane named "wds1", and the same for "its1".

I do not think that it matters whether the probability is high or low. It is something that really can happen, so users need to be told what will happen in that case.

Reply:
ok, so where do you want me to put the statement "if multiple CPs have the same name then only the last one found will be retained" in the help print out?

Reply:
I am not sure that it has to appear in the `--help` output, that is typically very succinct. Explaining this behavioral details in the chart doc file might be enough.